Spherical coordinate projections

lispandfound opened the pull request:
This PR is the first in a series that will refactor the codebase to align with EMOD3D internal coordinate projections. I have modelled this code directly on the EMOD3D codebase, so I am hoping to ensure the calculations exactly match the two going forward.

The projection used by internally EMOD3D is a [Gnomonic projection](https://en.wikipedia.org/wiki/Gnomonic_projection). This is equivalent to shining a light from the centre of the earth through a target point (the *origin*) onto its tangent plane. Areas and distances measured close to the origin are accurate, and the projection loses accuracy the farther you proceed from the centre. The projection is not valid past the same hemisphere as the origin point.

## How to use the new interface

Instead of reusing the old `xy2ll` and `ll2xy` code (which I intend to remove), the new code uses `pyproj` to construct the transformations directly. This means we get to maintain less code, and have stronger confidence in its correctness.

In addition, the interface to the transformations is now Pythonic, compare

``` python
amat, ainv = geo.gen_mat(mrot=0.0, mlon=172.0, mlat=-43.0)
lon, lat = -44.0, 172.0
geo.ll2xy(lon, lat) # 111.319439, 0
```

to the new interface

``` python
proj = SphericalProjection(mlat=-43.0, mlon=-172.0, mrot=0.0)
proj(-44.0, -172.0) # array([[  0.        , 111.33083037]])
proj.inverse(0.0, 111.33083037) # array([-44., -172.])
print(proj)
# SphericalProjection(mlat=-43.0, mlon=-172.0, mrot=0.0)
```

The new code is also vectorised.

joelridden reviewed (approved):
I assume some basic tests for this will come in some of the future PR's?

lispandfound commented:
This PR is now ready to review. Changes since the last draft:

1. We do not use pyproj. Tricky mathematical details of the projection mean that my old trick (transforming the coordinates after projection) was wrong. Instead this version is pretty faithful to original implementation plus extra numerical stability, better documentation of its qualities, and tests.
2. The documented projection changed. What I thought was a gnomonic projection is actually an azimuthal equidistant projection. This projection guarantees that:
   - Distances measured between the centre and any other point are correct.
   - All great-circles through the centre are projected as lines.
   It's important to recognise that makes this projection *not* a suitable replacement for NZTM because:
   - The projection makes no guarantees about distances, lines, angles or areas not near the centre.
   - Distortion is likely orders of magnitude higher than NZTM in the New Zealand region.

lispandfound commented on qcore/test/test_coordinates/test_coordinates.py:
```diff
@@ -0,0 +1,109 @@
+    # and longitude is 0 degrees (because longitude at the poles is undefined)
+    if np.isclose(np.abs(lat), 90):
+        assert pytest.approx(lat, abs=1e-5) == back[0]
+        assert pytest.approx(0.0, abs=1e-5) == 0.0
```
Oops!

lispandfound commented on qcore/coordinates.py:
```diff
@@ -216,3 +217,236 @@ def great_circle_bearing_to_nztm_bearing(
+                ],
+                [-cosA * sinT, -sinA * sinT, cosT],
+            ],
+            dtype="f",
```
Yeah that is probably fair.
